build/build.mk

```diff
@@ -87,23 +87,35 @@ define generatePlatformBuildTarget
 build-$(3)-$(1)-$(2):
   $(eval BINS_OUT_DIR_$(1)_$(2) := $(OUT_DIR)/$(1)_$(2)/$(BUILDTYPE_DIR))
 	@echo "$(ARROW) Building $(3) on $(1)/$(2) to $(BINS_OUT_DIR_$(1)_$(2))/$(3)$(BINARY_EXT)"
-	GOOS=$(1) GOARCH=$(2) go build \
+	cd $(4) && GOOS=$(1) GOARCH=$(2) go build \
 		-v \
 		-gcflags $(GCFLAGS) \
-		-ldflags=$(LDFLAGS) \
-		-o $(BINS_OUT_DIR_$(1)_$(2))/$(3)$(BINARY_EXT) \
-		$(4)/
+		-ldflags $(LDFLAGS) \
+		-o $(CURDIR)/$(BINS_OUT_DIR_$(1)_$(2))/$(3)$(BINARY_EXT)
 endef
 
 # defines a target for each binary
 GOOSES := darwin linux windows
 GOARCHES := amd64 arm arm64
-BINARIES := docgen rad applications-rp ucpd controller
-$(foreach ITEM,$(BINARIES),$(eval $(call generateBuildTarget,$(ITEM),./cmd/$(ITEM))))
-$(foreach ARCH,$(GOARCHES),$(foreach OS,$(GOOSES),$(foreach ITEM,$(BINARIES),$(eval $(call generatePlatformBuildTarget,$(OS),$(ARCH),$(ITEM),./cmd/$(ITEM))))))
+BINARIES := docgen:./cmd/docgen \
+	rad:./cmd/rad \
+	applications-rp:./cmd/applications-rp \
+	ucpd:./cmd/ucpd \
+	controller:./cmd/controller \
+	testrp:./test/testrp \
+	magpiego:./test/magpiego
```

Comment:
Some clarifying questions.. These are all good fodder for some improved comments for the existing logic.

So basically you're adding the test binaries and the paths to this data structure?

Then we use macros to parse these and generate a set of build targets dynamically.

Would be great to see an example. Each entry => these build targets.

Reply:
Added some comments. Let me know if those are clear. If not, I can add more.